book/src/rfcs/3-lookup.md

```diff
@@ -0,0 +1,291 @@
+
+This means two things for the lookup grand product argument:
+
+1. we cannot use the wrap around trick to make sure that the list is split in two correctly (enforced by $L_{n-1}(h_1(x) - h_2(g \cdot x)) = 0$ which is equivalent to $h_1(g^{n-1}) = h_2(1)$ in the plookup paper)
+2. we have even less space to store an entire query vector. Which is actually super correct, as the witness also has some zero-knowledge rows at the end that should not be part of the queries anyway.
+
+The first problem can be solved in two ways:
+
+* **Zig-zag technique**. By reorganizing $s$ to alternate its values between the columns. For example, $h_1 = (s_0, s_2, s_4, \cdots)$ and $h_2 = (s_1, s_3, s_5, \cdots)$ so that you can simply write the denominator of the grand product argument as 
+    $$(\gamma(1+\beta) + h_1(x) + \beta h_2(x))(\gamma(1+\beta)+ h_2(x) + \beta h_1(x \cdot g))$$
+    this is what the [plonkup](https://eprint.iacr.org/2022/086) paper does. 
+* **Snake technique**. by reorganizing $s$ as a snake. This is what is done in kimchi currently.
+
+The snake technique rearranges $s$ into the following shape:
+
+```
+    _   _
+ | | | | |
+ | | | | |
+ |_| |_| |
+```
+
+so that the denominator becomes the following equation:
+
+$$(\gamma(1+\beta) + h_1(x) + \beta h_1(x \cdot g))(\gamma(1+\beta)+ h_2(x \cdot g) + \beta h_2(x))$$
+
+and the snake doing a U-turn is constrained via something like
+
+$$L_{n-1} \cdot (h_1(x) - h_2(x)) = 0$$
+
+If there's an $h_3$ (because the table is very large, for example), then you'd have something like:
+
+$$(\gamma(1+\beta) + h_1(x) + \beta h_1(x \cdot g))(\gamma(1+\beta)+ h_2(x \cdot g) + \beta h_2(x))\color{green}{(\gamma(1+\beta)+ h_3(x) + \beta h_3(x \cdot g))}$$
+
+with the added U-turn constraint:
+
+$$L_{0} \cdot (h_2(x) - h_3(x)) = 0$$
+
+## Unsorted $t$ in $s$
+
+Note that at setup time, $t$ cannot be sorted as it is not combined yet. Since $s$ needs to be sorted by $t$ (in other words, not sorted, but sorted following the elements of $t$), there are two solutions:
+
+1. both the prover and the verifier can sort the combined $t$, so that $s$ can be sorted via the typical sorting algorithms
+2. the prover can sort $s$ by $t$, so that the verifier doesn't have to do any sorting and can just rely on the commitment of the columns of $t$ (which the prover can evaluate in the protocol).
+
+We do the second one, but there is an edge-case: the combined $t$ entries can repeat.
+For some $i, l$ such that $i \neq l$, we might have
+
+$$
+t_0[i] + j t_1[i] + j^2 t_2[i] = t_0[l] + j t_1[l] + j^2 t_2[l]
+$$
+
+For example, if $f = \{1, 2, 2, 3\}$ and $t = \{2, 1, 2, 3\}$, then $\text{sorted}(f, t) = \{2, 2, 1, 1, 2, 3\}$ would be one way of sorting things out. But $\text{sorted}(f, t) = \{2, 2, 2, 1, 1, 3\}$ would be incorrect.
```

Comment:
Here I really appreciate an example. Nonetheless, I got more questions from looking at it. I don't see the intuition of the "entries can repeat", and why the size of the sorted list has that size.

Reply:
we can talk about it more during our 1:1, but quickly:

* entries can repeat because there is the "combination of the columns with `j` to make the lookup table a single column table" that might lead to this. If we sort `t` then this is not a problem, but we want the verifier to avoid doing this work (due to the verifier circuit) and so we want to keep `t` unsorted
* the sorted vector is the concatenation of `f` and `t`. Let's ignore `t`. `f` is the 4 query vectors, where each query vector is constructed from the witness columns (and thus spanning all the rows of the domain). So each row is a mix of dummy queries and real queries so that there are 4 queries per row. So you can see `f` as 4 columns, where each column behaves like a gate